python/numpy.h

```diff
@@ -61,11 +63,38 @@ void copy_flattened_4d(const py::array_t<T> &data, View &&view) {
           *it = r(i, j, k, l);
 }
 
+template <class T> auto memory_begin_end(const py::buffer_info &info) {
+  auto *begin = static_cast<const T *>(info.ptr);
+  auto *end = static_cast<const T *>(info.ptr);
+  const auto &shape = info.shape;
+  const auto &strides = info.strides;
+  for (scipp::index i = 0; i < scipp::size(shape); ++i)
+    if (strides[i] < 0)
+      begin += shape[i] * strides[i];
+    else
+      end += shape[i] * strides[i];
+  return std::pair{begin, end};
+}
+
+template <class T, class View>
+bool memory_overlaps(const py::array_t<T> &data, const View &view) {
+  const auto &buffer_info = data.request();
+  const auto [data_begin, data_end] = memory_begin_end<T>(buffer_info);
+  const auto view_begin = &*view.begin();
+  const auto view_end = &*view.end();
+  // Note the use of std::less, pointer comparison with operator< may be
+  // undefined behavior with pointers from different arrays.
+  return std::less<const T *>()(data_begin, view_end) &&
+         std::greater_equal<const T *>()(data_end, view_begin);
+}
+
 template <class T, class View>
-void copy_flattened(const py::array_t<T> &data, View &&view) {
+void copy_flattened(py::array_t<T> data, View &&view) {
```

Comment:
Why not continue to take `data` as const ref?

Reply:
Because I may want to assign to it (a couple lines lower down).

Reply:
Discussed. `py::array<T>` is handle. This does not need to be changed.